docs: Improved the installation instructions of the Pleroma service

massimo-zaniboni opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

I installed the ``services.pleroma`` and so I acquired experience for improving the documentation.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

massimo-zaniboni committed:
docs: Improved the installation instructions of the Pleroma service

massimo-zaniboni committed:
Merge branch 'NixOS:master' into master

happysalada commented on nixos/modules/services/networking/pleroma.xml:
```diff
@@ -4,129 +4,194 @@
+<programlisting>
+services.postgresql = {
+  enable = true;
+  package = pkgs.postgresql_12;
```
perhaps default to 13 here ?

massimo-zaniboni replied:
done

happysalada commented on nixos/modules/services/networking/pleroma.xml:
```diff
@@ -4,129 +4,194 @@
+    config :pleroma, Pleroma.Web.Endpoint,
+      url: [host: "pleroma.example.net", scheme: "https", port: 443],
+      http: [ip: {127, 0, 0, 1}, port: 4000],
+      secret_key_base: "SECRET",
```
perhaps add these secrets as well to the secrets file ?

NinjaTrappeur replied:
Agree.

massimo-zaniboni replied:
To be fair these were not "secrets" but only place-holder strings, overriden by ``/var/lib/pleroma/secrets.exs`` real secrets. But you are right in any case, because it was not an optimal way to split secrets, so now I left secrets only on ``/var/lib/pleroma/secrets.exs``, and I used also a "fake-secret" for avoiding GitHub complaints about potentially exposed secrets. So I improved this part.

NinjaTrappeur replied:
I get it, but a reader might assume `SECRET` is a placeholder they need to fill with the actual values.

happysalada commented:
Thank you for this change! Documentation is critical and we need to improve it?

Could you not include the master merge commit ?

@petabyteboy @NinjaTrappeur 
Perhaps you have some additional thoughts ?

NinjaTrappeur commented on nixos/modules/services/networking/pleroma.xml:
```diff
@@ -4,129 +4,194 @@
+    config :pleroma, Pleroma.Repo,
+      adapter: Ecto.Adapters.Postgres,
+      username: "pleroma",
+      password: "SECRET",
```
Same. This should live in the path pointed by `secretConfigFile`, not in the store.

You can have a loot at the config of the NixOS test to see an example https://github.com/NixOS/nixpkgs/blob/3ab8ce12c2db31268f579c11727d9c63cfee2eee/nixos/tests/pleroma.nix#L123

massimo-zaniboni replied:
done

NinjaTrappeur commented on nixos/modules/services/networking/pleroma.xml:
```diff
@@ -4,129 +4,194 @@
+
+    "acmechallenge.example.net" = {
+      # Catchall vhost, will redirect users to HTTPS for all vhosts
+      serverAliases = [ "*.example.net" ];
```
Meh. Pleroma only requires a single vhost to operate. I don't see why you'd require a wildcard certificate just for this service.

Let's encourage users to use good security practices and only issue the certificate for the domain name we actually need.

massimo-zaniboni replied:
Ops! I miss interpreted the section of NixOS manual about ACME/Let's Encrypt https://nixos.org/manual/nixos/stable/index.html#module-security-acme-configuring . I thought this setting was mandatory. Instead it suffices a lot less!!!

I fixed this on my server, and tested creating new domains and it works. So I committed a simpler and safer version.

picnoir reviewed (commented):
Thanks a lot for this, this is a clear improvement!

Overall LGTM after we adress some comments I left in the review.

NinjaTrappeur commented:
CCing @yu-re-ka as they're also a pleroma maintainer.

happysalada commented:
Thank you for finding her, the maintainers list has yuka and I wasn't sure it was her.

massimo-zaniboni committed:
docs: Improved documentation of Pleroma service

Secrets are only in the private file, and in a draft format for avoiding GitHub
complaints.

Fixed Nginx configuration.

massimo-zaniboni committed:
Merge branch 'master' of github.com:massimo-zaniboni/nixpkgs

massimo-zaniboni commented:
> Thank you for this change! Documentation is critical and we need to improve it?
> Could you not include the master merge commit ?

I have no idea of how to avoid this now. Probably when someone merge, he can merge a squash version of my commits.

happysalada commented:
you can ammend commits in a PR with `git rebase -i --autosquash HEAD~2`
the i flag is for interactive and the autosquash is not mandatory.
When you have changed your branch history you can `git push --force`.
let me know if any of those doesn't make sense.
if any of those is a pain, I don't mind merging a squashed version, no problems on my side.

happysalada commented on nixos/modules/services/networking/pleroma.xml:
```diff
@@ -4,129 +4,174 @@
+  <para>In this configuration, Pleroma is listening only on the local port 4000. Nginx can be configured as a Reverse Proxy, for forwarding requests from public ports to the Pleroma service. This is an example of configuration, using
+<link xlink:href="https://letsencrypt.org/">Let's Encrypt</link> for the TLS certificates
+<programlisting>
+security.acme = {
```
if you are interested, instead of documenting those, a PR could be done about changing the nixos module and add a configuration `services.plerorma.nging.enable = true;` that would enable all of those settings.
this is exactly what is done here https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/services/web-apps/discourse.nix#L772
I understand that this is a little more involved, so no worries if you want to address this in a separate PR or not address this at all (both are completely fine with me).

massimo-zaniboni replied:
I have no time to code and test the ``services.pleroma.nginx.enable = true;`` but inspired from your link and from the configuration I'm using in production, I improved the Nginx configuration. In particular ``etags`` is very important for caching images and similar content, and it is probably mandatory on a Pleroma server.

NinjaTrappeur replied:
I personally dislike this kind of automagic module webserver configuration. From my experience, it tends to backfire quite quickly.

massimo-zaniboni committed:
docs: Improved the Nginx reverse proxy settings of Pleroma service

massimo-zaniboni commented:
> if any of those is a pain, I don't mind merging a squashed version, no problems on my side.

I prefer a squashed history on your side (i.e. NixOS repo), but a complete history of intermediate commits on the PR side (i.e. my temporary fork/branch). So, if it is fast for you, you can squash and merge the PR when you consider it good enough. I will mantain my history.

happysalada commented:
@massimo-zaniboni squashing here is no problem.

Can I request a final check from @NinjaTrappeur when you have a moment ?

NinjaTrappeur commented:
I'm busy ATM. I'll have a further look later in the week.

happysalada commented:
No rush, I'll wait for your approval before merging. Thank you for spending time on this!

picnoir reviewed (approved):
Massive thanks for your this doc overhaul @massimo-zaniboni, it was definitely needed!

I'm also very sorry about the latency here. It fell out of my brain at some point, may bad, my fault.

The Nginx conf turned out being alright.

I hope this won't frustrate you too much and turn you out of the NixOS project. I **really** hope to see you contribute to this project again in the future :)

I'm going to squash your commits a bit before merging them.

NinjaTrappeur commented:
Arf, Github is preventing me to push to your master branch (the one targeted by this PR).

I'm going to push your commit on a branch I can write on in a bit.

NinjaTrappeur commented:
I rebased everything for you there: https://github.com/NixOS/nixpkgs/pull/138281.

massimo-zaniboni commented:
> I hope this won't frustrate you too much and turn you out of the NixOS project. I really hope to see you contribute to this project again in the future :)

Don't worry: I have a lot of things queued myself, so I understand your situation!!! :-)

NinjaTrappeur commented:
Merged via a3d011eebcee9f70463a68c28b2252073336ed54

NinjaTrappeur commented:
And now it's merged, I realize I botched the commit message by forgetting to remove squashed messages :disappointed:
